checks.d/kubernetes.py

```diff
@@ -235,25 +245,72 @@ def _update_container_metrics(self, instance, subcontainer, kube_labels):
                               sum(float(net[x]) for x in NET_ERRORS),
                               tags)
 
-    def _retrieve_metrics(self, url):
-        return retrieve_json(url)
+        return tags
 
-    def _update_metrics(self, instance):
-        pods_list = self.kubeutil.retrieve_pods_list()
-        metrics = self._retrieve_metrics(self.kubeutil.metrics_url)
+    def _update_metrics(self, instance, pods_list):
+        metrics = self.kubeutil.retrieve_metrics()
 
         excluded_labels = instance.get('excluded_labels')
         kube_labels = self.kubeutil.extract_kube_labels(pods_list, excluded_keys=excluded_labels)
 
         if not metrics:
             raise Exception('No metrics retrieved cmd=%s' % self.metrics_cmd)
 
+        # container metrics from Cadvisor
+        container_tags = {}
         for subcontainer in metrics:
+            c_id = subcontainer.get('id')
+            try:
+                tags = self._update_container_metrics(instance, subcontainer, kube_labels)
+                if c_id:
+                    container_tags[c_id] = tags
+                # also store tags for aliases
+                for alias in subcontainer.get('aliases', []):
+                    container_tags[alias] = tags
+            except Exception, e:
+                self.log.error("Unable to collect metrics for container: {0} ({1}".format(c_id, e))
+
+        # container metrics from kubernetes API: limits and requests
+        for pod in pods_list['items']:
             try:
-                self._update_container_metrics(instance, subcontainer, kube_labels)
-            except Exception as e:
-                self.log.error("Unable to collect metrics for container: {0} ({1}".format(
-                    subcontainer.get('name'), e))
+                containers = pod['spec']['containers']
+                name2id = {}
+                for cs in pod['status'].get('containerStatuses', []):
+                    c_id = cs.get('containerID', '').split('//')[-1]
+                    name = cs.get('name')
+                    if name:
+                        name2id[name] = c_id
+            except KeyError:
+                self.log.debug("Pod %s does not have containers specs, skipping...", pod['metadata'].get('name'))
+                continue
+
+            for container in containers:
+                c_name = container.get('name')
+                _tags = container_tags.get(name2id.get(c_name), [])
+
+                # limits
+                try:
+                    for limit, value_str in container['resources']['limits'].iteritems():
+                        values = [float(s) for s in re.findall(r'[-+]?\d+[\.]?\d*', value_str)]
```

Comment:
maybe you should compile this regex, it's used in a loop here, and also for requests.